### PR TITLE
Add more details when reset password required error is received, Fixes AB#3084603

### PR DESCRIPTION
**Pull request by nilo-ms**
When reset password is required during a call to /token, we need some native auth specific handling to inform developers on how to handle this error.

[AB#3084603](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3084603)

### CODE REVIEW

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java [hunk 1]
```diff
@@ -236,6 +236,19 @@ public static ServiceException getExceptionFromTokenErrorResponse(@NonNull final
                     developerDescription,
                     apiError
             );
+        } else if (isNativeAuthenticationResetPasswordRequiredError(errorResponse)) {
+            ServiceException apiError = new ServiceException(
+                    errorResponse.getError(),
+                    errorResponse.getErrorDescription(),
+                    null);
+
+            String developerDescription = "Password change is required, which can't be fulfilled as part of this flow."+
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java]: nit: maybe mention the word "user" here, to make it clear this is about a user's password and an action the user has to complete.

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt [hunk 2]
```diff
@@ -81,6 +82,16 @@ class SignInTokenApiResponse(
                             correlationId = correlationId
                         )
                     }
+                    errorCodes[0].isPasswordChangeRequired() -> {
+                        val customDescription = "Password change is required, which can't be fulfilled as part of this flow. " +
```
Comment (SammyO) [inline common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt]: Is this necessary? You're already adding this information in `ExceptionAdapter`. Also, I'd recommend keeping the original API error description and not altering them. Developer should have a way to access to the raw API response.

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java (hunk 1 above)
Comment (nilo-ms) [reply common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java]: Thanks. Should this message be "User password change is required, which can't be..."?

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt (hunk 2 above)
Comment (nilo-ms) [reply common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt]: I think the "ExceptionAdapter" class handle the error when we're refreshing the access token, while in this case we handle the scenario of obtaining a new token. Correct me if I'm wrong.
About the error description, I'm adding the description from API at the end of the custom message. Do you suggest adding the custom string in a different place?

File: common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java (hunk 1 above)
Comment (SammyO) [reply common4j/src/main/com/microsoft/identity/common/java/controllers/ExceptionAdapter.java]: Sounds good to me.

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt (hunk 2 above)
Comment (SammyO) [reply common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt]: > Correct me if I'm wrong.

You're probably right, but I'm not 100% sure - you'll have to run tests.

**Review by GBakolas — APPROVED** [review]

File: common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt (hunk 2 above)
Comment (nilo-ms) [reply common4j/src/main/com/microsoft/identity/common/java/nativeauth/providers/responses/signin/SignInTokenApiResponse.kt]: Ok, I did a test, and this code is executed during the signIn. So, my assumption was correct

**Review by SammyO — APPROVED** [review]

**Review by spetrescu84 — APPROVED** [review]

**Pull request merged**